File: src/visual-graph/builder/describe-subgraph.test.ts
import { describe, expect, test } from "vitest";

import { SCOPE_TYPE, type ScopeType } from "../../analyzer/scope-type.js";
import { asScopeId } from "../../ir/serialized/scope-id.js";
import type { SerializedScope } from "../../ir/serialized/serialized-scope.js";
import type { SerializedVariable } from "../../ir/serialized/serialized-variable.js";
import { asVariableId } from "../../ir/serialized/variable-id.js";
import { AST_TYPE } from "../../parser/ast-type.js";
import { asFilledString } from "../../util/filled-string.js";
import { DIRECTION } from "../direction.js";
import { SUBGRAPH_KIND } from "../subgraph-kind.js";
import { VISUAL_ELEMENT_TYPE } from "../visual-element-type.js";
import type { VisualSubgraph } from "../visual-subgraph.js";
import { describeSubgraph } from "./describe-subgraph.js";
import { baseCaseClauseBlockContext } from "./testing/make-block-context.js";
import { baseScope } from "./testing/make-scope.js";
import { baseVariable } from "./testing/make-variable.js";
import { span } from "./testing/span.js";

describe("describeSubgraph", () => {
  test("function subgraph returns kind=function with owner metadata", () => {
    const fnScope = {
      ...baseScope(),
      id: asScopeId("fnScope"),
      type: SCOPE_TYPE.Function,
      block: {
        type: AST_TYPE.FunctionDeclaration,
        span: span(0, 5),
        endSpan: span(50, 10),
      },
    } as const satisfies SerializedScope;
    const owner = {
      ...baseVariable(),
      id: asVariableId("ownerVar"),
      name: asFilledString("myFn"),
      identifiers: [span(0, 5)],
    } as const satisfies SerializedVariable;
    const owners = new Map([["fnScope", "ownerVar"]]);
    const variables = new Map([["ownerVar", owner]]);

    const sg = describeSubgraph(fnScope, owners, variables);

    expect(sg).toMatchObject({
      type: VISUAL_ELEMENT_TYPE.Subgraph,
      id: "s_fnScope",
      kind: SUBGRAPH_KIND.Function,
      direction: DIRECTION.RL,
      ownerNodeId: "n_ownerVar",
      ownerName: "myFn",
      line: 5,
      endLine: 10,
      elements: [],
    });
  });

  test("function subgraph falls back to scope.block.span.line when owner has no identifiers", () => {
    const fnScope = {
      ...baseScope(),
      id: asScopeId("fn"),
      type: SCOPE_TYPE.Function,
      block: {
        type: AST_TYPE.FunctionDeclaration,
        span: span(0, 7),
        endSpan: span(20, 9),
      },
    } as const satisfies SerializedScope;
    const owner = {
      ...baseVariable(),
      id: asVariableId("o"),
      name: asFilledString("f"),
      identifiers: [],
    } as const satisfies SerializedVariable;
    const sg = describeSubgraph(
      fnScope,
      new Map([["fn", "o"]]),
      new Map([["o", owner]]),
    );
    expect(sg.line).toEqual(7);
  });

  test("function subgraph without an owner var renders as anonymous (ownerNodeId null, ownerName empty)", () => {
    const scope = {
      ...baseScope(),
      id: asScopeId("fn"),
      type: SCOPE_TYPE.Function,
    } as const satisfies SerializedScope;
    const sg = describeSubgraph(scope, new Map(), new Map());
    expect(sg.kind).toEqual(SUBGRAPH_KIND.Function);
    if (sg.kind === SUBGRAPH_KIND.Function) {
      expect(sg.ownerNodeId).toEqual(null);
      expect(sg.ownerName).toEqual("");
    }
  });

  test.each<{
    name: string;
    type: ScopeType;
    expectedKind: VisualSubgraph["kind"];
  }>([
    {
      name: asFilledString("for"),
      type: SCOPE_TYPE.For,
      expectedKind: SUBGRAPH_KIND.For,
    },
    {
      name: asFilledString("catch"),
      type: SCOPE_TYPE.Catch,
      expectedKind: SUBGRAPH_KIND.Catch,
    },
    {
      name: asFilledString("switch"),
      type: SCOPE_TYPE.Switch,
      expectedKind: SUBGRAPH_KIND.Switch,
    },
  ])(
    "control subgraph for scope type $name -> kind=$expectedKind",
    ({ type, expectedKind }) => {
      const scope = {
        ...baseScope(),
        id: asScopeId("ctrl"),
        type,
        block: {
          type: AST_TYPE.BlockStatement,
          span: span(0, 1),
          endSpan: span(10, 3),
        },
      } as const satisfies SerializedScope;
      const sg = describeSubgraph(scope, new Map(), new Map());
      expect(sg.kind).toEqual(expectedKind);
      expect(sg.id).toEqual("s_ctrl");
      expect(sg.line).toEqual(1);
      expect(sg.endLine).toEqual(3);
    },
  );

  test("case subgraph captures caseTest from blockContext", () => {
    const scope = {
      ...baseScope(),
      id: asScopeId("case1"),
      type: SCOPE_TYPE.Block,
      block: {
        type: AST_TYPE.BlockStatement,
        span: span(0, 1),
        endSpan: span(10, 2),
      },
      blockContext: {
        ...baseCaseClauseBlockContext(),
        caseTest: asFilledString("x === 1"),
      },
    } as const satisfies SerializedScope;
    const sg = describeSubgraph(scope, new Map(), new Map());
    expect(sg.kind).toEqual(SUBGRAPH_KIND.Case);
    if (sg.kind !== SUBGRAPH_KIND.Case) {
      throw new Error("expected case");
    }
    expect(sg.caseTest).toEqual("x === 1");
  });

  test("case subgraph keeps caseTest null when the case clause has no test (default)", () => {
    const scope = {
      ...baseScope(),
      id: asScopeId("case-default"),
      type: SCOPE_TYPE.Block,
      blockContext: baseCaseClauseBlockContext(),
    } as const satisfies SerializedScope;
    const sg = describeSubgraph(scope, new Map(), new Map());
    if (sg.kind !== SUBGRAPH_KIND.Case) {
      throw new Error("expected case");
    }
    expect(sg.caseTest).toEqual(null);
  });

  test("plain block scope renders as the generic 'block' subgraph", () => {
    const scope = {
      ...baseScope(),
      id: asScopeId("plain"),
      type: SCOPE_TYPE.Block,
    } as const satisfies SerializedScope;
    const sg = describeSubgraph(scope, new Map(), new Map());
    expect(sg.kind).toEqual(SUBGRAPH_KIND.Block);
  });

  test("class scope with a ClassName binding picks up the inner identifier as className", () => {
    const classScope = {
      ...baseScope(),
      id: asScopeId("clsScope"),
      type: SCOPE_TYPE.Class,
      block: {
        type: AST_TYPE.ClassExpression,
        span: span(0, 2),
        endSpan: span(30, 4),
      },
      variables: [asVariableId("innerNameVar")],
    } as const satisfies SerializedScope;
    const inner = {
      ...baseVariable(),
      id: asVariableId("innerNameVar"),
      name: asFilledString("Foo"),
      identifiers: [span(0, 2)],
    } as const satisfies SerializedVariable;
    const sg = describeSubgraph(
      classScope,
      new Map(),
      new Map([["innerNameVar", inner]]),
    );
    expect(sg).toMatchObject({
      type: VISUAL_ELEMENT_TYPE.Subgraph,
      id: "s_clsScope",
      kind: SUBGRAPH_KIND.Class,
      direction: DIRECTION.RL,
      className: "Foo",
      line: 2,
      endLine: 4,
      elements: [],
    });
  });

  test("class scope with no variables (anonymous ClassExpression) yields className=null", () => {
    const classScope = {
      ...baseScope(),
      id: asScopeId("anon"),
      type: SCOPE_TYPE.Class,
      block: {
        type: AST_TYPE.ClassExpression,
        span: span(0, 1),
        endSpan: span(10, 1),
      },
      variables: [],
    } as const satisfies SerializedScope;
    const sg = describeSubgraph(classScope, new Map(), new Map());
    expect(sg.kind).toEqual(SUBGRAPH_KIND.Class);
    if (sg.kind !== SUBGRAPH_KIND.Class) {
      throw new Error("expected class");
    }
    expect(sg.className).toEqual(null);
  });

  test("throws when scope is neither a function subgraph nor a control kind (e.g. module / global)", () => {
    const scope = {
      ...baseScope(),
      id: asScopeId("mod"),
      type: SCOPE_TYPE.Module,
    } as const satisfies SerializedScope;
    expect(() => describeSubgraph(scope, new Map(), new Map())).toThrow();
  });
});

File: crates/unsnarl-visual-graph/src/builder/describe_subgraph_test.rs
//! Sibling tests for [`describe_subgraph`].

use std::collections::HashMap;

use unsnarl_ir::scope_type::ScopeType;
use unsnarl_ir::serialized::serialized_scope::SerializedBlock;
use unsnarl_ir::serialized::SerializedVariable;
use unsnarl_oxc_parity::AstType;

use super::describe_subgraph;
use crate::builder::builder_fixtures::{
    base_serialized_scope, base_serialized_variable, case_clause_block_context, scope_id,
    span_offset_line, variable_id,
};
use crate::direction::Direction;
use crate::visual_subgraph::{
    ControlExtras, ControlSubgraphKind, OwnedExtras, OwnedSubgraphKind, VisualSubgraph,
};

fn block(
    r#type: AstType,
    span_offset: u32,
    span_line: u32,
    end_offset: u32,
    end_line: u32,
) -> SerializedBlock {
    SerializedBlock {
        r#type,
        span: span_offset_line(span_offset, span_line),
        end_span: span_offset_line(end_offset, end_line),
    }
}

#[test]
fn function_subgraph_returns_kind_function_with_owner_metadata() {
    let mut scope = base_serialized_scope("fnScope");
    scope.r#type = ScopeType::Function;
    scope.block = block(AstType::FunctionDeclaration, 0, 5, 50, 10);

    let owner = SerializedVariable::new(
        variable_id("ownerVar"),
        "myFn".to_string(),
        scope_id("s"),
        vec![span_offset_line(0, 5)],
        Vec::new(),
        Vec::new(),
    );
    let mut owners: HashMap<String, String> = HashMap::new();
    owners.insert("fnScope".to_string(), "ownerVar".to_string());
    let mut variables: HashMap<&str, &SerializedVariable> = HashMap::new();
    variables.insert("ownerVar", &owner);

    let sg = describe_subgraph(
        &scope,
        &owners,
        &variables,
        &unsnarl_ir::primitive::SourceIndex::build(""),
    );
    let VisualSubgraph::Owned(s) = sg else {
        panic!("expected owned subgraph");
    };
    assert_eq!(s.id, "s_fnScope");
    assert!(matches!(s.kind, OwnedSubgraphKind::Function));
    assert!(matches!(s.direction, Direction::RL));
    assert_eq!(s.line, 5);
    assert_eq!(s.end_line, Some(10));
    assert!(s.elements.is_empty());
    let OwnedExtras::Function {
        owner_node_id,
        owner_name,
    } = s.extras
    else {
        panic!("expected Function extras");
    };
    assert_eq!(owner_node_id.as_deref(), Some("n_ownerVar"));
    assert_eq!(owner_name, "myFn");
}

#[test]
fn function_subgraph_falls_back_to_scope_block_span_line_when_owner_has_no_identifiers() {
    let mut scope = base_serialized_scope("fn");
    scope.r#type = ScopeType::Function;
    scope.block = block(AstType::FunctionDeclaration, 0, 7, 20, 9);
    let owner = SerializedVariable::new(
        variable_id("o"),
        "f".to_string(),
        scope_id("s"),
        Vec::new(),
        Vec::new(),
        Vec::new(),
    );
    let mut owners: HashMap<String, String> = HashMap::new();
    owners.insert("fn".to_string(), "o".to_string());
    let mut variables: HashMap<&str, &SerializedVariable> = HashMap::new();
    variables.insert("o", &owner);
    let VisualSubgraph::Owned(s) = describe_subgraph(
        &scope,
        &owners,
        &variables,
        &unsnarl_ir::primitive::SourceIndex::build(""),
    ) else {
        panic!("expected owned");
    };
    assert_eq!(s.line, 7);
}

#[test]
fn function_subgraph_without_owner_var_renders_anonymous() {
    let mut scope = base_serialized_scope("fn");
    scope.r#type = ScopeType::Function;
    let VisualSubgraph::Owned(s) = describe_subgraph(
        &scope,
        &HashMap::new(),
        &HashMap::new(),
        &unsnarl_ir::primitive::SourceIndex::build(""),
    ) else {
        panic!("expected owned");
    };
    assert!(matches!(s.kind, OwnedSubgraphKind::Function));
    let OwnedExtras::Function {
        owner_node_id,
        owner_name,
    } = s.extras
    else {
        panic!("expected Function extras");
    };
    assert_eq!(owner_node_id, None);
    assert_eq!(owner_name, "");
}

#[test]
fn control_subgraph_for() {
    let mut scope = base_serialized_scope("ctrl");
    scope.r#type = ScopeType::For;
    scope.block = block(AstType::BlockStatement, 0, 1, 10, 3);
    let VisualSubgraph::Control(s) = describe_subgraph(
        &scope,
        &HashMap::new(),
        &HashMap::new(),
        &unsnarl_ir::primitive::SourceIndex::build(""),
    ) else {
        panic!("expected control");
    };
    assert!(matches!(s.kind, ControlSubgraphKind::For));
    assert_eq!(s.id, "s_ctrl");
    assert_eq!(s.line, 1);
    assert_eq!(s.end_line, Some(3));
}

#[test]
fn control_subgraph_catch() {
    let mut scope = base_serialized_scope("ctrl");
    scope.r#type = ScopeType::Catch;
    scope.block = block(AstType::BlockStatement, 0, 1, 10, 3);
    let VisualSubgraph::Control(s) = describe_subgraph(
        &scope,
        &HashMap::new(),
        &HashMap::new(),
        &unsnarl_ir::primitive::SourceIndex::build(""),
    ) else {
        panic!("expected control");
    };
    assert!(matches!(s.kind, ControlSubgraphKind::Catch));
}

#[test]
fn control_subgraph_switch() {
    let mut scope = base_serialized_scope("ctrl");
    scope.r#type = ScopeType::Switch;
    scope.block = block(AstType::BlockStatement, 0, 1, 10, 3);
    let VisualSubgraph::Control(s) = describe_subgraph(
        &scope,
        &HashMap::new(),
        &HashMap::new(),
        &unsnarl_ir::primitive::SourceIndex::build(""),
    ) else {
        panic!("expected control");
    };
    assert!(matches!(s.kind, ControlSubgraphKind::Switch));
}

#[test]
fn case_subgraph_captures_case_test_from_block_context() {
    let mut scope = base_serialized_scope("case1");
    scope.r#type = ScopeType::Block;
    scope.block = block(AstType::BlockStatement, 0, 1, 10, 2);
    scope.block_context = Some(case_clause_block_context(
        AstType::SwitchStatement,
        "cases",
        0,
        Some("x === 1"),
    ));
    let VisualSubgraph::Control(s) = describe_subgraph(
        &scope,
        &HashMap::new(),
        &HashMap::new(),
        &unsnarl_ir::primitive::SourceIndex::build(""),
    ) else {
        panic!("expected control");
    };
    assert!(matches!(s.kind, ControlSubgraphKind::Case));
    let ControlExtras::Case { case_test } = &s.extras else {
        panic!("expected Case extras");
    };
    assert_eq!(case_test.as_deref(), Some("x === 1"));
}

#[test]
fn case_subgraph_keeps_case_test_null_when_default() {
    let mut scope = base_serialized_scope("case-default");
    scope.r#type = ScopeType::Block;
    scope.block_context = Some(case_clause_block_context(
        AstType::SwitchStatement,
        "cases",
        0,
        None,
    ));
    let VisualSubgraph::Control(s) = describe_subgraph(
        &scope,
        &HashMap::new(),
        &HashMap::new(),
        &unsnarl_ir::primitive::SourceIndex::build(""),
    ) else {
        panic!("expected control");
    };
    let ControlExtras::Case { case_test } = &s.extras else {
        panic!("expected Case extras");
    };
    assert_eq!(case_test.as_deref(), None);
}

#[test]
fn plain_block_scope_renders_as_generic_block() {
    let scope = base_serialized_scope("plain");
    let VisualSubgraph::Control(s) = describe_subgraph(
        &scope,
        &HashMap::new(),
        &HashMap::new(),
        &unsnarl_ir::primitive::SourceIndex::build(""),
    ) else {
        panic!("expected control");
    };
    assert!(matches!(s.kind, ControlSubgraphKind::Block));
}

#[test]
fn class_scope_with_class_name_binding_picks_inner_identifier() {
    let mut scope = base_serialized_scope("clsScope");
    scope.r#type = ScopeType::Class;
    scope.block = block(AstType::ClassExpression, 0, 2, 30, 4);
    scope.variables = vec![variable_id("innerNameVar")];
    let inner = SerializedVariable::new(
        variable_id("innerNameVar"),
        "Foo".to_string(),
        scope_id("s"),
        vec![span_offset_line(0, 2)],
        Vec::new(),
        Vec::new(),
    );
    let _ = base_serialized_variable; // suppress unused warning in scope-local imports
    let mut variables: HashMap<&str, &SerializedVariable> = HashMap::new();
    variables.insert("innerNameVar", &inner);

    let VisualSubgraph::Owned(s) = describe_subgraph(
        &scope,
        &HashMap::new(),
        &variables,
        &unsnarl_ir::primitive::SourceIndex::build(""),
    ) else {
        panic!("expected owned");
    };
    assert_eq!(s.id, "s_clsScope");
    assert!(matches!(s.kind, OwnedSubgraphKind::Class));
    assert!(matches!(s.direction, Direction::RL));
    assert_eq!(s.line, 2);
    assert_eq!(s.end_line, Some(4));
    let OwnedExtras::Class { class_name } = s.extras else {
        panic!("expected Class extras");
    };
    assert_eq!(class_name.as_deref(), Some("Foo"));
}

#[test]
fn class_scope_with_no_variables_yields_class_name_null() {
    let mut scope = base_serialized_scope("anon");
    scope.r#type = ScopeType::Class;
    scope.block = block(AstType::ClassExpression, 0, 1, 10, 1);
    scope.variables = Vec::new();
    let VisualSubgraph::Owned(s) = describe_subgraph(
        &scope,
        &HashMap::new(),
        &HashMap::new(),
        &unsnarl_ir::primitive::SourceIndex::build(""),
    ) else {
        panic!("expected owned");
    };
    let OwnedExtras::Class { class_name } = s.extras else {
        panic!("expected Class extras");
    };
    assert_eq!(class_name, None);
}

#[test]
#[should_panic]
fn panics_when_scope_is_module() {
    let mut scope = base_serialized_scope("mod");
    scope.r#type = ScopeType::Module;
    let _ = describe_subgraph(
        &scope,
        &HashMap::new(),
        &HashMap::new(),
        &unsnarl_ir::primitive::SourceIndex::build(""),
    );
}
